### ISSUE TITLE
Annotation Clustering Identifier

### ISSUE THREAD

**Issue by RyanCopley** [mapbox/DEPRECATED-mapbox-ios-sdk#439]
It would be handy to be able to cluster similar identifiers, and not cluster different ones. Currently, it's only possible to disable clustering for the ones you don't want and only enable one set to cluster. I'm going to look into how to do this, but I thought I would add it as an issue first.

My idea is to add an identifier to the annotations, and only allow clustering between similar identifiers.

If anyone has any insight of this, feel free to let me know, I'm hoping to do this within the next few days

**Comment by incanus** [comment]
This would likely involve multiple quad trees per map view instead of just one to track these cluster sets separately. That whole part of the code could use some attention so I'm curious to see what you think of this as you poke around.

**Comment by RyanCopley** [comment]
Currently this is what I've been trying: https://gist.github.com/RyanCopley/11231560

(And I've added annotationIdentifier to the RMAnnotation class), I'm working with the Cocoapods version so I don't think I can commit and push, I'll have to copy.

Althought, I'll say that this is not working exactly as I was hoping. I'm having a hard time analyzing that function, but we'll see what comes up.

**Comment by RyanCopley** [comment]
https://github.com/RyanCopley/mapbox-ios-sdk/commit/23f6087da0b487328a218b52ffb43cadba2d473c

This seems to work, could you take a look and see if there is anything obviously wrong with it?

To setup, setup multiple different annotations and give them different clusterIdentifiers (NSStrings, this could be anything though, ints, etc, FWIW, that might be something to think about)

I also stopped it from clustering when zoomed in all the way. That could optionally be removed.

If you do not set a clusterIdentifier, it uses the default one (Which would group all annotations that do not have a unique identifier)

Things that might be worrisome / deserve a discussion: 
1) Speed implication of the predicate searches
2) Over-removing annotations
3) This patch vs potentially refactoring (Or using different quad trees per identifier)
4) The averaging of coordinates seems very incorrect. Not sure if my code broke this or if it was already broken.

**Comment by RyanCopley** [comment]
The someArray in 

```
    RMQuadTree -addAnnotationsInBoundingBox:(RMProjectedRect)aBoundingBox
                     toMutableArray:(NSMutableArray *)someArray
           createClusterAnnotations:(BOOL)createClusterAnnotations
           withProjectedClusterSize:(RMProjectedSize)clusterSize
      andProjectedClusterMarkerSize:(RMProjectedSize)clusterMarkerSize
                  findGravityCenter:(BOOL)findGravityCenter
```

confuses me-- What is this someArray for?

**Comment by incanus** [comment]
> This would likely involve multiple quad trees per map view instead of just one to track these cluster sets separately.

This was me (not) thinking quickly -- you might also try the single quad tree but qualifying by your identifier(s). That is, markers don't change location based on identifier, so the single quad tree structure would still apply, but you'd just possibly draw differently based on identifier.

**Comment by incanus** [comment]
> What is this someArray for? 

And this one I don't recall offhand -- would have to dig.

**Comment by RyanCopley** [comment]
I found an issue in my implementation, this one works better: https://github.com/RyanCopley/mapbox-ios-sdk/commit/dbdef8ec57f2a1f246bb70ad2f2588e203d99c8f

I was accidentally re-filtering the results over and over again. This is obviously not right :)
I'll roll these into one commit once I feel like this is good enough

After some more research, this still isn't entirely correct. I'm getting cluster annotations with no clusteredAnnotations in them...

**Comment by RyanCopley** [comment]
I analyzed this a bit more and I determined some major flaws in my current implementation, I think I have the right idea of how to fix this :+1:

**Comment by RyanCopley** [comment]
https://github.com/RyanCopley/mapbox-ios-sdk/commit/7d9ab7c8c8fe1bfa8fb48b263b051f980a28ff1a

This seems to work great. The commit might need cleaned up however. If we can agree this is a good solution, I can re-write it to be a bit neater.

**Comment by RyanCopley** [comment]
There is a need for some @synchronized blocks around some of the code I didn't address... Some race conditions were identified :(

**Comment by mohpor** [comment]
Any news on this? I need this concept implemented asap.
Is there anything I can do to make it happen?

**Comment by RyanCopley** [comment]
I have it sort of working, but there are some glitches I'm still working out. I'm getting ghost annotations :(

**Comment by mohpor** [comment]
I'm not sure how can I help, but if you are getting ghost annotations, I think you should review your `mapView:(RMMapView *)mapView layerForAnnotation:(RMAnnotation *)annotation`'s implementation to find out what are these ghost annotations. Let me know if you've already explored this. seems like we can solve this soon.
Good luck

**Comment by RyanCopley** [comment]
Lol yes, I'm well aware of that delegate method. The issue lies inside my changes to the map box SDK-- I'll make them public on Monday when I get in the office

**Comment by mohpor** [comment]
:D i didn't mean it like that!Thanks a lot
—

**Comment by mohpor** [comment]
Did you commit anything @RyanCopley ?

**Comment by mohpor** [comment]
@RyanCopley , I'm still waiting for your commits, let me know if you've got time to do it.
Cheers

**Comment by RyanCopley** [comment]
Sorry... I'm doing it now

**Comment by RyanCopley** [comment]
https://github.com/RyanCopley/mapbox-ios-sdk 

I'm still combing through it to make sure I didn't include anything stupid in that commit... but that's my live working copy so far. I've disabled annotation clustering completely for now in my app though

Edit: There's a lot of stupid stuff in there... but the idea is there still. I have a lot of small changes that i've made that you might not want

To use it, create your annotation and apply this:

```
annotation.clusteringEnabled = YES;
annotation.clusterIdentifier = @"Some obscure identifier for this cluster group";
```

**Comment by mohpor** [comment]
Thanks @RyanCopley,
I've had to disable clustering too, mainly because the clustering is not usable as is off the box. I needed something to do more accurate clustering for me, like an identifier, or a key to let mapview know what markers should it cluster... so I disabled clustering altogether for the time being and pushed the task down my tasks stack for a while, but I'm at that point now which I need to concretely implement this feature, whether MapBox iOS SDK can deliver it or not.
I wish both of us luck and I hope @incanus would chip in and check on this as frequent as reasonably possible. I can see that he's truly busy improving the SDK (especially rendering, multithreaded tile loading, etc. ...) but I'm not sure he's working on other clustering concepts as well.

Many thanks and I hope I'll be of help to you as well.
Cheers

**Comment by RyanCopley** [comment]
@mohpor: I'm able to dedicate some time to this. If you'd like to collaborate and I could help explain my changes, we might be able to find the issues if you're free? Let me know and we can exchange some form of chat details

**Comment by mohpor** [comment]
that's awesome @RyanCopley,
We have a giant timezone difference but I think we can work through it.
I have updated my githup profile so that you can reach me via email and exchange chat info.

**Comment by incanus** [comment]
Any forward progress on this @RyanCopley @mohpor?

**Comment by RyanCopley** [comment]
My implementation is still slightly too buggy :(

**Comment by mohpor** [comment]
@incanus 
This is in my long list of tasks for my undergoing project, but as the whole idea is not fully disclosed with me, I'm now awaiting more information. I might need to have conditional clustering based on zoom level, area, etc. therefore, I'm going to pause and wait for the features to come in. I was thinking about having a dive into QuadTree but my hands are tied to something right now.
Looking forward to use the GL by the way. I hope raster tile sources (custom TMS) get supported asap so that I could start migrating before it's too late for me.

**Comment by incanus** [comment]
>  I hope raster tile sources (custom TMS) get supported asap so that I could start migrating before it's too late for me.

These are supported now. Check out something like this, which is raster-based: 

https://github.com/mapbox/mapbox-gl-styles/blob/mb-pages/styles/satellite.json

You'd just want to use a standard tile template URL for the `url` value.

**Comment by buttcmd** [comment]
@incanus have you merged this into your code or something similar to cluster only a type on annotations on your standard (no gl) sdk?

It looks very interesting but i would like to remain on the trunk

Thank you

**Comment by incanus** [comment]
No, there is no distinction between clustered and non-clustered annotations right now in this project.

**Comment by buttcmd** [comment]
Ok. Thanks for your answer

**Comment by iv-mexx** [comment]
> This was me (not) thinking quickly -- you might also try the single quad tree but qualifying by your identifier(s). That is, markers don't change location based on identifier, so the single quad tree structure would still apply, but you'd just possibly draw differently based on identifier.

#630 is my attempt to implement clustering by clusteringIdentifier using the single quad tree
